tsfc/impero_utils.py

```diff
@@ -27,6 +28,90 @@
 ImperoC = collections.namedtuple('ImperoC', ['tree', 'temporaries', 'declare', 'indices'])
 
 
+class NoopError(Exception):
+    """No operations in the kernel."""
+    pass
+
+
+def compile_gem(return_variables, expressions, prefix_ordering, remove_zeros=False, coffee_licm=False):
+    """Compiles GEM to Impero.
+
+    :arg return_variables: return variables for each root (type: GEM expressions)
+    :arg expressions: multi-root expression DAG (type: GEM expressions)
+    :arg prefix_ordering: outermost loop indices
+    :arg remove_zeros: remove zero assignment to return variables
+    :arg coffee_licm: trust COFFEE to do loop invariant code motion
+    """
+    return_variables = optimise.remove_componenttensors(return_variables)
+    expressions = optimise.remove_componenttensors(expressions)
+
+    # Remove zeros
+    if remove_zeros:
+        rv = []
+        es = []
+        for var, expr in zip(return_variables, expressions):
+            if not isinstance(expr, gem.Zero):
+                rv.append(var)
+                es.append(expr)
+        return_variables, expressions = rv, es
+
+    # Collect indices in a deterministic order
+    indices = []
+    for node in traversal(expressions):
+        if isinstance(node, gem.Indexed):
+            indices.extend(node.multiindex)
+    # The next two lines remove duplicate elements from the list, but
+    # preserve the ordering, i.e. all elements will appear only once,
+    # in the order of their first occurance in the original list.
+    _, unique_indices = numpy.unique(indices, return_index=True)
+    indices = numpy.asarray(indices)[numpy.sort(unique_indices)]
+
+    # Build ordered index map
+    index_ordering = make_prefix_ordering(indices, prefix_ordering)
+    apply_ordering = make_index_orderer(index_ordering)
+
+    get_indices = lambda expr: apply_ordering(expr.free_indices)
+
+    # Build operation ordering
+    ops = scheduling.emit_operations(zip(return_variables, expressions), get_indices)
+
+    # Empty kernel
+    if len(ops) == 0:
+        raise NoopError()
```

Comment:
I don't really like exceptions for control flow, but I don't see a better way of doing this!